File: AdmitadSDKDemo/ObjCDemoApp/CartVC.m
//
//  ViewController.m
//  fff
//
//  Created by Dmitry Cherednikov on 08/11/2017.
//  Copyright © 2017 Tachos. All rights reserved.
//

#import "CartVC.h"
#import "AppDelegate.h"

@import AdmitadSDK;


typedef NS_ENUM(NSInteger, CartActionType) {
    CartActionTypeConfirm,
    CartActionTypePay,
    CartActionTypeNone
};

@interface CartVC () <UITableViewDataSource> {

    IBOutlet __weak UIBarButtonItem *rightNavigationBarItem;
    IBOutlet __weak UITableView *tableView;
    IBOutlet __weak UIActivityIndicatorView *spinner;

    CartActionType currentAction;
    NSString *currencyCode;

    AdmitadTracker *admitadTracker;

    NSArray<AdmitadOrderItem *> *items;
    NSArray<NSNumber *> *prices;
}

@end

@implementation CartVC

// MARK:- UIViewController Life Cycle 
- (void)viewDidLoad {

    [super viewDidLoad];

    currentAction = CartActionTypeConfirm;
    currencyCode = @"RUB";
    admitadTracker = [AdmitadTracker sharedInstance];

    tableView.dataSource = self;
    self.title = @"Cart";

    [self initItemsAndPrices];
}

- (void)initItemsAndPrices {

    AdmitadOrderItem *item1 = [[AdmitadOrderItem alloc] initWithName:@"Phone"
                                                            quantity:1];
    AdmitadOrderItem *item2 = [[AdmitadOrderItem alloc] initWithName:@"Phone Charger"
                                                            quantity:3];
    AdmitadOrderItem *item3 = [[AdmitadOrderItem alloc] initWithName:@"TV"
                                                            quantity:2];
    AdmitadOrderItem *item4 = [[AdmitadOrderItem alloc] initWithName:@"USB Flash Drive"
                                                            quantity:2];

    items = @[item1, item2, item3, item4];
    prices = @[@3000, @200, @45000, @800];
}

// MARK:- UITableViewDataSource
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {

    return items.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {

    NSString * reuseId = @"cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:reuseId];
    if (!cell) {

        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleValue1
                                      reuseIdentifier:reuseId];
    }

    AdmitadOrderItem *item = items[indexPath.row];

    cell.textLabel.text = item.name;
    cell.detailTextLabel.text = [self detailTextForRow: indexPath.row];

    return cell;
}

// MARK:- Action
- (IBAction)rightNavigationBarItemPressed {

    switch (currentAction) {
        case CartActionTypeConfirm:
            [self trackConfirmedPurchaseEvent];
            break;
        case CartActionTypePay:
            [self trackPaidOrderEvent];
            break;
        case CartActionTypeNone:
            break;
    }
}

// MARK:- Events Tracking
- (void)trackConfirmedPurchaseEvent {

    [spinner startAnimating];

    AppDelegate *appDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    [admitadTracker trackConfirmedPurchaseEventWithOrder:[self buildOrder] channel:nil completion:^(AdmitadError *error) {

        [spinner stopAnimating];
        if (!error) {

            currentAction = CartActionTypePay;
            rightNavigationBarItem.title = @"Pay";
            [appDelegate showAlertWithMessage:@"Confirmed Purchase event is successfully tracked"];
        }
        else {

            [appDelegate showAlertWithMessage:@"Error tracking Confirmed Purchase event"];
        }
    }];
}

- (void)trackPaidOrderEvent {

    [spinner startAnimating];

    AppDelegate *appDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    [admitadTracker trackPaidOrderEventWithOrder:[self buildOrder] channel:nil completion:^(AdmitadError *error) {

        [spinner stopAnimating];
        if (!error) {

            currentAction = CartActionTypeNone;
            rightNavigationBarItem.enabled = NO;
            [appDelegate showAlertWithMessage:@"Paid Order event is successfully tracked"];
        }
        else {

            [appDelegate showAlertWithMessage:@"Error tracking Paid Order event"];
        }
    }];
}

// MARK:- Utils
- (NSString *)detailTextForRow: (NSInteger)row {

    AdmitadOrderItem *item = items[row];
    NSString *priceString = [NSString stringWithFormat:@"%.0f", prices[row].floatValue];

    return [NSString stringWithFormat:@"%d x %@ %@",
            (int)item.quantity, priceString,
            currencyCode];
}

- (AdmitadOrder *)buildOrder {

    NSDictionary<NSString *, NSString *> *userInfo = @{@"country": @"Russia",
                                                       @"payment_method": @"PayPal"};

    return [[AdmitadOrder alloc] initWithId:@"123456789"
                                 totalPrice:[self orderPrice]
                                 currencyCode:currencyCode
                                 items:items
                                 userInfo:userInfo
                                 tarifCode:@"1"
                                 promocode:@""
                                ];
}

- (NSString *)orderPrice {
    CGFloat sum = 0.f;
    for (NSInteger i = 0; i < 4; ++i) {

        sum += items[i].quantity * prices[i].floatValue;
    }

    return [NSString stringWithFormat: @"%f", sum];
}

@end

File: AdmitadSDKDemo/SwiftDemoApp/CartVC.swift
//
//  ViewController.swift
//  DeeplinkSDKTest
//
//  Created by Dmitry Cherednikov on 26.09.17.
//  Copyright © 2017 tachos. All rights reserved.
//

import UIKit
import AdmitadSDK

enum CartActionType: String {
    case confirm = "confirm"
    case pay = "pay"
    case none = "none"
}

class CartVC: UIViewController, UITableViewDataSource {
    @IBOutlet weak var rightNavigationBarItem: UIBarButtonItem!
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var spinner: UIActivityIndicatorView!

    private var currentAction = CartActionType.confirm
    private let currencyCode = "RUB"
    private let admitadTracker = AdmitadTracker.sharedInstance

    private let items: [AdmitadOrderItem] = {
        let item1 = AdmitadOrderItem(name: "Phone")
        let item2 = AdmitadOrderItem(name: "Phone Charger",
                                     quantity: 3)
        let item3 = AdmitadOrderItem(name: "TV",
                                     quantity: 2)
        let item4 = AdmitadOrderItem(name: "USB Flash Drive",
                                     quantity: 4)
        return [item1, item2, item3, item4]
    }()

    private let prices = [3000, 200, 45000, 800]
}

// MARK: UIViewController lifecycle
extension CartVC {
    override func viewDidLoad() {
        super.viewDidLoad()

        tableView.dataSource = self

        title = "Cart"
    }
}

// MARK: UITableViewDataSource
extension CartVC {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return items.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let reuseId = "cell"
        var cell = tableView.dequeueReusableCell(withIdentifier: reuseId)
        if cell == nil {
            cell = UITableViewCell(style: .value1,
                                   reuseIdentifier: reuseId)
        }
        let item = items[indexPath.row]
        
        cell!.textLabel?.text = item.name
        cell!.detailTextLabel?.text = detailText(forRow: indexPath.row)

        return cell!
    }
}

// MARK: actions
private extension CartVC {
    @IBAction func rightNavigationBarItemPressed(_ sender: Any) {
        switch currentAction {
        case .confirm:
            trackConfirmedPurchaseEvent()
        case .pay:
            trackPaidOrderEvent()
        case .none:
            break
        }
    }

    func trackConfirmedPurchaseEvent() {
        spinner.startAnimating()

        admitadTracker.trackConfirmedPurchaseEvent(order: buildOrder())
        {   [weak self] (error) in
            self?.spinner.stopAnimating()

            if error != nil {
                return
            }

            self?.currentAction = .pay
            self?.rightNavigationBarItem.title = CartActionType.pay.rawValue
        }
    }

    func trackPaidOrderEvent() {
        spinner.startAnimating()

        admitadTracker.trackPaidOrderEvent(order: buildOrder())
        {   [weak self] (error) in
            self?.spinner.stopAnimating()

            if error != nil {
                return
            }

            self?.currentAction = .none
            self?.rightNavigationBarItem.isEnabled = false
        }
    }
}

// MARK: - utility methods
private extension CartVC {
    func detailText(forRow row: Int) -> String {
        let item = items[row]
        let price = prices[row]
        return String(item.quantity) + " x " + String(price) +
            " " + currencyCode
    }
}

// MARK: - order building
private extension CartVC {
    func buildOrder() -> AdmitadOrder {
        return AdmitadOrder(id: "123456789",
                            totalPrice: orderPrice(),
                            currencyCode: currencyCode,
                            items: items,
                            userInfo: ["country": "Russia",
                                       "payment_method": "PayPal"], tarifCode: "123", promocode: "")
    }

    func orderPrice() -> String {
        var sum = 0
        for i in 0...3 {
            sum += items[i].quantity * prices[i]
        }
        return String(describing: sum)
    }
}
